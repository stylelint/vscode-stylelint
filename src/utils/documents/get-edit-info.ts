import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import type { LintDiagnostics } from '../stylelint/index';
import type LSP from 'vscode-languageserver-protocol';

/**
 * Get the edit info for a diagnostic.
 * @param document The document to get edit info for.
 * @param diagnostic The diagnostic to get edit info for.
 */
export function getEditInfo(
	document: TextDocument,
	diagnostic: LSP.Diagnostic,
	lintResult:
		| (LintDiagnostics & {
				version: number;
		  })
		| undefined,
): { label: string; edit: TextEdit } | undefined {
	if (!lintResult || document.version !== lintResult.version) {
		return undefined;
	}

	const warning = lintResult.getWarning?.(diagnostic);

	if (!warning) {
		return undefined;
	}

	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore -- (TS2353) `fix` property is available since v16.15.
	const edit = warning.fix;

	if (!edit) {
		return undefined;
	}

	return {
		label: `Fix this ${warning.rule} problem`,
		edit: {
			newText: edit.text,
			range: {
				start: document.positionAt(edit.range[0]),
				end: document.positionAt(edit.range[1]),
			},
		},
	};
}
