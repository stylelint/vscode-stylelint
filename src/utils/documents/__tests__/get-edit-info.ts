import { TextDocument } from 'vscode-languageserver-textdocument';
import * as LSP from 'vscode-languageserver-protocol';

import { getEditInfo } from '../get-edit-info';

const DOCUMENT_VERSION = 1;
const createDocument = (code: string) =>
	TextDocument.create('file:///path/test.css', 'css', DOCUMENT_VERSION, code);

describe('getEditInfo', () => {
	it('should construct and return an EditInfo', async () => {
		const document = createDocument('a {color: #000;}');

		const warning = {
			rule: 'color-hex-length',
			text: `Expected "#000" to be "#000000" (color-hex-length)`,
			line: 1,
			column: 10,
			endLine: 1,
			endColumn: 14,
			severity: 'error' as const,
			fix: {
				range: [12, 13] as [number, number],
				text: '0000',
			},
		};
		const diagnostic = {
			range: {
				start: document.positionAt(10),
				end: document.positionAt(14),
			},
			message: `Expected "#000" to be "#000000" (color-hex-length)`,
			source: 'stylelint',
			severity: LSP.DiagnosticSeverity.Error,
			code: 'color-hex-length',
		};

		const editInfo = getEditInfo(document, diagnostic, {
			version: DOCUMENT_VERSION,
			diagnostics: [diagnostic],
			getWarning: () => warning,
		});

		expect(editInfo).toMatchSnapshot();
	});

	it('should return undefined if the document does not have a lint result', async () => {
		const document = createDocument('a {color: #000;}');
		const diagnostic = {
			range: {
				start: document.positionAt(10),
				end: document.positionAt(14),
			},
			message: `Expected "#000" to be "#000000" (color-hex-length)`,
			source: 'stylelint',
			severity: LSP.DiagnosticSeverity.Error,
			code: 'color-hex-length',
		};

		const editInfo = getEditInfo(document, diagnostic, undefined);

		expect(editInfo).toBeUndefined();
	});

	it('should return undefined if the document version does not match', async () => {
		const document = createDocument('a {color: #000;}');
		const warning = {
			rule: 'color-hex-length',
			text: `Expected "#000" to be "#000000" (color-hex-length)`,
			line: 1,
			column: 10,
			endLine: 1,
			endColumn: 14,
			severity: 'error' as const,
			fix: {
				range: [12, 13] as [number, number],
				text: '0000',
			},
		};
		const diagnostic = {
			range: {
				start: document.positionAt(10),
				end: document.positionAt(14),
			},
			message: `Expected "#000" to be "#000000" (color-hex-length)`,
			source: 'stylelint',
			severity: LSP.DiagnosticSeverity.Error,
			code: 'color-hex-length',
		};

		const editInfo = getEditInfo(document, diagnostic, {
			version: 2,
			diagnostics: [diagnostic],
			getWarning: () => warning,
		});

		expect(editInfo).toBeUndefined();
	});

	it('should return undefined if the diagnostic has no warning', async () => {
		const document = createDocument('a {color: #000;}');
		const diagnostic = {
			range: {
				start: document.positionAt(10),
				end: document.positionAt(14),
			},
			message: `Expected "#000" to be "#000000" (color-hex-length)`,
			source: 'stylelint',
			severity: LSP.DiagnosticSeverity.Error,
			code: 'color-hex-length',
		};

		const editInfo = getEditInfo(document, diagnostic, {
			version: DOCUMENT_VERSION,
			diagnostics: [diagnostic],
			getWarning: () => null,
		});

		expect(editInfo).toBeUndefined();
	});

	it('should return undefined if the warning has no fix', async () => {
		const document = createDocument('a {color: #000;}');

		const warning = {
			rule: 'color-hex-length',
			text: `Expected "#000" to be "#000000" (color-hex-length)`,
			line: 1,
			column: 10,
			endLine: 1,
			endColumn: 14,
			severity: 'error' as const,
		};
		const diagnostic = {
			range: {
				start: document.positionAt(10),
				end: document.positionAt(14),
			},
			message: `Expected "#000" to be "#000000" (color-hex-length)`,
			source: 'stylelint',
			severity: LSP.DiagnosticSeverity.Error,
			code: 'color-hex-length',
		};

		const editInfo = getEditInfo(document, diagnostic, {
			version: DOCUMENT_VERSION,
			diagnostics: [diagnostic],
			getWarning: () => warning,
		});

		expect(editInfo).toBeUndefined();
	});
});
