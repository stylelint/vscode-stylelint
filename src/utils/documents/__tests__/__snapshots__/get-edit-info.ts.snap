// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEditInfo should construct and return an EditInfo 1`] = `
{
  "edit": {
    "newText": "0000",
    "range": {
      "end": {
        "character": 13,
        "line": 0,
      },
      "start": {
        "character": 12,
        "line": 0,
      },
    },
  },
  "label": "Fix this color-hex-length problem",
}
`;
